streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1681,6 +1688,26 @@ class ZStream[-R, +E, +A](val process: ZManaged[R, E, Pull[R, E, A]]) extends Se
       } yield pull
     }
 
+  /**
+   * Takes all elements of the stream until the specified predicate evaluates
+   * to `true`.
+   */
+  def takeUntil(pred: A => Boolean): ZStream[R, E, A] =
+    ZStream[R, E, A] {
+      for {
+        as            <- self.process
+        keepTakingRef <- Ref.make(true).toManaged_
+        pull = keepTakingRef.get.flatMap { p =>
```

Comment:
does this action need to be atomic? if yes then it might be useful to use `modify` instead

Reply:
I thought we were okay because we were inside the managed from `process`  but I'm not 100% sure. @iravid or @vasilmkd?

Reply:
All bets are off with regard to interruption of a `Pull`.

If you can do things atomically that's better, but basically everything breaks if you interrupt a `Pull`.

Reply:
Okay. It looks like we use the same `get` and `set` pattern in the existing implementation of `dropWhile`.

Reply:
We could, rewrite all cases where we can do things atomically. But yes, it's as grim as @iravid puts it. And at that point, I'm not sure if the extra work is worth it.